go2ts.go

```
@@ -0,0 +1,311 @@
+	// typeName is the TypeScript type, such as "string", or "map", or "SomeInterfaceName".
+	typeName string
+
+	// keyType is the type of the key if this tsType is a map, such as "string" or "number".
+	keyType string
+
+	// interfaceName is the name of the TypeScript interface if the tsType is
+	// "interface".
+	interfaceName string
+
+	canBeNull bool
+	subType   *tsType
+}
+
+// String returns the tsType formatted as TypeScript.
+func (s tsType) String() string {
+	var ret string
+	switch s.typeName {
+	case "array":
+		ret = s.subType.String() + "[]"
+	case "map":
+		ret = fmt.Sprintf("{ [key: %s]: %s }", s.keyType, s.subType.String())
+	case "interface":
+		ret = s.interfaceName
+	default:
+		ret = s.typeName
+	}
+
+	return ret
+}
+
+// fieldRep represents one field in a struct.
+type fieldRep struct {
+	tsType     tsType
+	name       string
+	isOptional bool
+}
+
+func (f fieldRep) String() string {
+	optional := ""
+	if f.isOptional {
+		optional = "?"
+	}
+
+	canBeNull := ""
+	if f.tsType.canBeNull {
+		canBeNull = " | null"
+	}
+
+	return fmt.Sprintf("\t%s%s: %s%s;", f.name, optional, f.tsType.String(), canBeNull)
+}
+
+func (f *fieldRep) initializeFromReflectStructField(structField reflect.StructField) {
+	jsonTag := strings.Split(structField.Tag.Get("json"), ",")
+
+	f.name = structField.Name
+	if len(jsonTag) > 0 && jsonTag[0] != "" {
+		f.name = jsonTag[0]
+	}
+	f.isOptional = len(jsonTag) > 1 && jsonTag[1] == "omitempty"
+}
+
+func isTime(t reflect.Type) bool {
+	return t.Name() == "Time" && t.PkgPath() == "time"
+}
+
+// structRep represents a single Go struct.
+type structRep struct {
+	// Name is the TypeScript interface Name in the generated output.
+	Name   string
+	Fields []fieldRep
+}
+
+var structRepTemplate = template.Must(template.New("").Parse(`
+export interface {{ .Name }} {
+{{ range .Fields -}}
+	{{- . }}
+{{ end -}}
+}
+`))
+
+func (s *structRep) render(w io.Writer) (err error) {
+	return structRepTemplate.Execute(w, s)
```

IIUC this error only happens in the presence of a programming error, i.e. if structRepTemplate and structRep disagree on how things are named. This is independent of the user's input.

If my understanding is correct, then I would recommend not piping through this error and failing loudly with a panic instead (with comments explaining this reasoning), as there is no meaningful way to recover.

---

Done.